Accept None as Value for Limit Parameter for jobs()

andre-a-alves opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fixes a bug and allows `None` to be passed as a value for the `limit` parameter when calling jobs().


### Details and comments
The docstring for the jobs() method identifies `limit=None` as the way to pull all jobs. However, the method would fail with that limit, because at one point a value needed to be added to the `limit` value.
Closes #586

kt474 commented on test/integration/test_ibm_job.py:
```diff
@@ -139,6 +139,14 @@ def test_retrieve_jobs(self):
+            limit=None,
+        )
+        self.assertGreaterEqual(len(job_list), 0)
+
```
thanks for fixing this @andre-a-alves! 

Some accounts like the one used in the github CI have many jobs (my account has 60k) so this test case will just time out. I can't think of another way to test this so I'd be ok with just removing this

andre-a-alves replied:
I completely understand you point, and having all of 87 jobs in my account, I clearly didn't think big enough 😂. However, I do not think it is the best idea to leave out the test entirely, since not testing for this case is likely how the bug was introduced in the first place.

I went ahead an added some unit tests that mock `IBMBackendService` and some of its components and methods. The mocks provide dummy data to be processed by the `jobs()` and `_get_jobs()` methods to ensure they return the expected number of jobs. Actually, implementing the unit tests helped me find some errors in my original solution that have also been corrected.

If this is not a satisfactory solution, please let me know, and I'll see if I can come up with an alternative.

andre-a-alves commented:
I also went ahead and made another minor tweak to `jobs()`. `limit` was being stored in `original_limit` before the ending while loop, but it was never modified. So I just removed `original_limit`.

andre-a-alves commented:
I was wondering if there is any feedback on whether this PR is ok as it exists right now, something needs to be changed, or something is unclear/being discussed. Is anything waiting on me?

merav-aharoni commented on qiskit_ibm_provider/ibm_backend_service.py:
```diff
@@ -322,8 +322,8 @@ def jobs(
         job_list = []
-        original_limit = limit
         while True:
+            added_jobs = 0
```
Remove - see my next comment.

andre-a-alves replied:
changes applied

merav-aharoni commented on qiskit_ibm_provider/ibm_backend_service.py:
```diff
@@ -362,9 +362,10 @@ def jobs(
                         continue
                     job_list.append(job)
-                    if len(job_list) == original_limit:
+                    added_jobs += 1
```
Remove - it is incorrect to skip only the jobs that were added. You must skip all the jobs that were returned.

andre-a-alves replied:
changes applied

merav-aharoni commented on qiskit_ibm_provider/ibm_backend_service.py:
```diff
@@ -362,9 +362,10 @@ def jobs(
+                    if limit and len(job_list) == limit:
                         return job_list
-            skip += limit
+            skip += limit if limit else added_jobs
```
```suggestion
            skip += len(job_responses)
```

andre-a-alves replied:
changes applied

merav-aharoni reviewed (commented):
Sorry it took so long to get to this PR.
In addition to the fix I pointed out in the comments, we'd prefer not to add a test with so much overhead - new decorator, new file. I suggest adding something simpler, like the following test to `test/integration/test/test_ibm_provider.py`:
```
    def test_jobs_filters(self):
        circuit = QuantumCircuit(1)
        circuit.h(0)
        num_jobs = 60 # at least 50, because that is page size
        provider = IBMProvider()
        backend_name = "ibmq_qasm_simulator"
        backend = provider.get_backend(backend_name)
        start = datetime.now()
        for _ in range(num_jobs):
            job = backend.run(circuit)
        end = datetime.now()
        recent_jobs_limit = provider.jobs(backend_name=backend_name,
                                          limit=100,
                                          start_datetime=start,
                                          end_datetime=end)
        recent_jobs_no_limit = provider.jobs(backend_name=backend_name,
                                             limit=None,
                                             start_datetime=start,
                                             end_datetime=end)
        self.assertEqual(len(recent_jobs_limit), num_jobs)
        self.assertEqual(len(recent_jobs_no_limit), num_jobs)
```

andre-a-alves commented:
@merav-aharoni Thank you for the comments. I am leaving for a trip, but I will take care of this when I return at the end of the month.

andre-a-alves commented:
Thank you for the suggested integration test. I modified it slightly and implemented it, and I also removed the unit tests and their overhead.

I took a look at the failed CI action, and it appears to be in a portion of the code that I did not touch.

If there are any more changes required, please feel free to reach out :-)

merav-aharoni commented:
@andre-a-alves - Looks great now! thank you for all your hard work!
The PR must pass all tests before we can approve it. I agree that it seems unrelated to your work. I suggest you rerun the tests (I am not sure how - easiest is to make some tiny commit, or update when `main` changes).

andre-a-alves commented:
@merav-aharoni I dug a bit into the problem and opened Issue #687

kt474 reviewed (approved):
Thanks @andre-a-alves!
